web/pages/api/v0/bet/index.ts

```diff
@@ -1,24 +1,308 @@
+    ) {
+      throw new APIError(400, 'Trade too large for current liquidity pool.')
+    }
+
+    const betDoc = contractDoc.collection('bets').doc()
+    trans.create(betDoc, {
+      id: betDoc.id,
+      userId: user.id,
+      userAvatarUrl: user.avatarUrl,
+      userUsername: user.username,
+      userName: user.name,
+      ...newBet,
+    })
+    log(`Created new bet document for ${user.username} - auth ${userId}.`)
+
+    if (makers) {
+      updateMakers(makers, betDoc.id, contractDoc, trans)
+    }
+    if (ordersToCancel) {
+      for (const bet of ordersToCancel) {
+        trans.update(contractDoc.collection('bets').doc(bet.id), {
+          isCancelled: true,
+        })
+      }
+    }
+
+    const balanceChange =
+      newBet.amount !== 0
+        ? // quick bet
+          newBet.amount + FLAT_TRADE_FEE
+        : // limit order
+          FLAT_TRADE_FEE
+
+    trans.update(userDoc, { balance: FieldValue.increment(-balanceChange) })
+    log(`Updated user ${user.username} balance - auth ${userId}.`)
+
+    if (newBet.amount !== 0) {
+      trans.update(
+        contractDoc,
+        removeUndefinedProps({
+          pool: newPool,
+          p: newP,
+          totalShares: newTotalShares,
+          totalBets: newTotalBets,
+          totalLiquidity: newTotalLiquidity,
+          collectedFees: addObjects(newBet.fees, collectedFees),
+          volume: volume + newBet.amount,
+        })
+      )
+      log(`Updated contract ${contract.slug} properties - auth ${userId}.`)
+    }
+
+    return { contract, betId: betDoc.id, makers, newBet }
   })
-  try {
-    const backendRes = await fetchBackend(req, 'placebet')
-    await forwardResponse(res, backendRes)
-  } catch (err) {
-    console.error('Error talking to cloud function: ', err)
-    res.status(500).json({ message: 'Error communicating with backend.' })
+
+  log(`Main transaction finished - auth ${userId}.`)
+
+  const { contract, newBet, makers } = result
+  const { mechanism } = contract
+
+  if (
+    (mechanism === 'cpmm-1' || mechanism === 'cpmm-2') &&
+    newBet.amount !== 0
+  ) {
+    const userIds = uniq([
+      userId,
+      ...(makers ?? []).map((maker) => maker.bet.userId),
+    ])
+    await Promise.all(userIds.map((userId) => redeemShares(userId, contract)))
+    log(`Share redemption transaction finished - auth ${userId}.`)
+  }
+
+  return { betId: result.betId }
+})
+
+export default placeBet
+
+const bodySchema = z.object({
+  contractId: z.string(),
+  amount: z.number().gte(1),
+})
+
+const binarySchema = z.object({
+  outcome: z.enum(['YES', 'NO']),
+  limitProb: z.number().gte(0).lte(1).optional(),
+})
+
+const freeResponseSchema = z.object({
+  outcome: z.string(),
+  shortSell: z.boolean().optional(),
+})
+
+const numericSchema = z.object({
+  outcome: z.string(),
+  value: z.number(),
+})
+
+export type PlaceBetParams = z.infer<
+  typeof bodySchema &
+    (typeof binarySchema | typeof freeResponseSchema | typeof numericSchema)
+>
```

Comment:
I guess we could dream of a more precise type here, since this isn't capturing the correlation between the body content and the type of the contract-type-specific parameters.